### ISSUE TITLE
Error when trying to use Sphinx to write documentation of code using pydantic

### ISSUE THREAD

**Issue by cyrilbouvier** [samuelcolvin/pydantic#2500]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.1
            pydantic compiled: True
                 install path: /home/cyril/src/sage-9.2/local/lib/python3.8/site-packages/pydantic
               python version: 3.8.5 (default, Jan  4 2021, 11:37:14)  [GCC 8.3.0]
                     platform: Linux-4.19.0-12-amd64-x86_64-with-glibc2.2.5
     optional deps. installed: ['typing-extensions']

```

I have a very simple package using pydantic, the file `mypackage/__init__.py` contains

```py
import sys
import pydantic

print (f'# python version: {sys.version}')
print (f'# pydantic version: {pydantic.VERSION}')

@pydantic.dataclasses.dataclass
class Foo:
  name: str

# also fails with a BaseModel
#class Foo (pydantic.BaseModel):
#  name: str

F = Foo (name='TEST')
```

When executing the following script
```py
import mypackage

print (mypackage.F)
```
I obtain the following expected output

```
# python version: 3.8.5 (default, Jan  4 2021, 11:37:14) 
[GCC 8.3.0]
# pydantic version: 1.8.1
Foo(name='TEST')
```

If I try to use Sphinx to build documentation for this package, I obtain an error.

I used the following command to invoke sphinx:
```
sphinx-build -b html doc/source doc/build/html
```

The file `doc/source/index.rst` contains:

```
documentation
=============

.. automodule:: mypackage
```

The error message that Sphinx returns is:
```
WARNING: autodoc: failed to import module 'mypackage'; the following exception was raised:
Traceback (most recent call last):
  File "/home/cyril/src/sage-9.2/local/lib/python3.8/site-packages/sphinx/ext/autodoc/importer.py", line 32, in import_module
    return importlib.import_module(modname)
  File "/home/cyril/src/sage-9.2/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/cyril/tmp/python/ScalMul/mypackage/__init__.py", line 15, in <module>
    F = Foo (name='TEST')
  File "<string>", line 4, in __init__
  File "pydantic/dataclasses.py", line 99, in pydantic.dataclasses._generate_pydantic_post_init._pydantic_post_init
pydantic.error_wrappers.ValidationError: 1 validation error for Foo
name
  str_validator() takes exactly one argument (5 given) (type=type_error)
```

From what I could track from the pydantic source code, the problem comes from the last list of the above `__init__.py` file: `F = Foo(name='TEST')`. Following all the function calls, it comes down to the method `_apply_validators` of the `ModelField` class. The list `validators` only contains one validator: `<cyfunction str_validator at 0x.....>`. Looking at the definition of this function in `validators.py`, it appears that it take only 1 argument (the value of the field) but inside the function `_apply_validators` (at line 913) it is called with 5 arguments
```py
v = validator(cls, v, values, self, self.model_config)
```
As the function is printed as a `cyfunction` I suspect that the code is not supposed to call the function `str_validator` from the python code but from one of the C libraries `*.cpython-38-x86_64-linux-gnu.so` of the pydantic directory.

Note: it seems that Sphinx use imporlib.import_module to import the module, not a classic `import`.

# TL;DR

Why, when importing from a python script, the correct `str_validator` function is called, but when Sphinx tries to import the package, it calls the python function `str_validator` ?

**Comment by cyrilbouvier** [comment]
Sorry for the noise, I finally found the origin of the problem: in the `conf.py` file generated by sphinx-quickstart, the `functools.wraps` function is modified. This is the root of the problem that I observed but this is not related to pydantic.
